### PR TITLE
[csharp-netcore] Fixed handling of nullable enum values

### LINKED ISSUE #5442: [BUG] [csharp-netcore] Invalid expressions are generated using ??
#### Bug Report Checklist

- [x] Have you provided a full/minimal spec to reproduce the issue?
- [x] Have you validated the input using an OpenAPI validator ([example](https://apidevtools.org/swagger-parser/online/))?
- [x] What's the version of OpenAPI Generator used?
- [x] Have you search for related issues/PRs?
- [x] What's the actual output vs expected output?
- [ ] [Optional] Bounty to sponsor the fix ([example](https://www.bountysource.com/issues/66123212-javascript-client-produces-a-wrong-object-for-a-string-enum-type-that-is-used-with-ref))

##### Description
The csharp-netcore generator generates invalid C# if an enum is set as required in a definition. With the enum property name present in the "required" object inside the definition of `ObjectCreateArgs`, the C# below is generated as a constructor for `ObjectCreateArgs`:

```csharp
public ObjectCreateArgs(ObjectType objectType = default(ObjectType))
{
    // to ensure "objectType" is required (not null)
    this.ObjectType = objectType ?? throw new ArgumentNullException("snip");;
}
```

This is invalid C# because `objectType` is not nullable, therefore the null coalescing operator can't be used with it.

##### openapi-generator version
4.2.3

##### OpenAPI declaration file content or url
[Click here](https://github.com/OpenAPITools/openapi-generator/files/4252477/swagger.txt) (rename extension to .json)

##### Command line used for generation
```
java -jar openapi-generator-cli.jar generate -i swagger.json -g csharp-netcore -o test-csharp
```

##### Steps to reproduce
Run the above command.

##### Suggest a fix
Enums are not nullable unless declared with `?` like other types, so they should be treated as such.

### PR DESCRIPTION
**Pull request by shibayan**
Fixes an issue where the enum type is not properly treated as nullable.

Fixed #5442 

### PR checklist

- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] If contributing template-only or documentation-only changes which will change sample output, [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) before.
- [x] Run the shell script(s) under `./bin/` (or Windows batch scripts under`.\bin\windows`) to update Petstore samples related to your fix. This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit, and these must match the expectations made by your contribution. You only need to run `./bin/{LANG}-petstore.sh`, `./bin/openapi3/{LANG}-petstore.sh` if updating the code or mustache templates for a language (`{LANG}`) (e.g. php, ruby, python, etc).
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master`, `4.3.x`, `5.0.x`. Default: `master`.
- [x] Copy the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) to review the pull request if your PR is targeting a particular programming language.

### CODE REVIEW

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractCSharpCodegen.java [hunk 1]
```diff
@@ -603,6 +604,25 @@ protected void updateValueTypeProperty(Map<String, Object> models) {
...
+     * @param models list of all models
+     */
+    protected void updateNullableTypeProperty(Map<String, Object> models) {
+        for (Map.Entry<String, Object> entry : models.entrySet()) {
+            String openAPIName = entry.getKey();
+            CodegenModel model = ModelUtils.getModelByName(openAPIName, models);
+            if (model != null) {
+                for (CodegenProperty var : model.vars) {
+                    if (!var.isContainer && (nullableType.contains(var.dataType) || var.isEnum)) {
+                        var.vendorExtensions.put("x-csharp-value-type", true);
```
Comment (shibayan) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractCSharpCodegen.java:619]: Only at this timing `isEnum` is set to the correct value

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractCSharpCodegen.java [hunk 2]
```diff
@@ -692,6 +719,37 @@ protected void processOperation(CodegenOperation operation) {
...
+                }
+            }
+
+            if (model != null) {
+                // Effectively mark enum models as enums and non-nullable
+                if (model.isEnum) {
+                    parameter.isEnum = true;
+                    parameter.allowableValues = model.allowableValues;
+                    parameter.isPrimitiveType = true;
+                    parameter.vendorExtensions.put("x-csharp-value-type", true);
```
Comment (shibayan) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractCSharpCodegen.java:739]: Only an enum with type generation can be determined here.

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/CSharpNetCoreClientCodegen.java [hunk 3]
```diff
@@ -423,10 +423,6 @@ public void postProcessModelProperty(CodegenModel model, CodegenProperty propert
         postProcessEmitDefaultValue(property.vendorExtensions);
 
         super.postProcessModelProperty(model, property);
-
-        if (!property.isContainer && (nullableType.contains(property.dataType) || property.isEnum)) {
-            property.vendorExtensions.put("x-csharp-value-type", true);
```
Comment (shibayan) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/CSharpNetCoreClientCodegen.java:428]: `isEnum` is always `false` at this timing

File: modules/openapi-generator/src/main/resources/csharp-netcore/api.mustache [hunk 4]
```diff
@@ -256,36 +256,48 @@ namespace {{packageName}}.{{apiPackage}}
             if (localVarAccept != null) localVarRequestOptions.HeaderParameters.Add("Accept", localVarAccept);
 
             {{#pathParams}}
-            {{^vendorExtensions.x-csharp-value-type}}
+            {{#required}}
+            localVarRequestOptions.PathParameters.Add("{{baseName}}", {{packageName}}.Client.ClientUtils.ParameterToString({{paramName}})); // path parameter
+            {{/required}}
```
Comment (shibayan) [inline modules/openapi-generator/src/main/resources/csharp-netcore/api.mustache:261]: It has been confirmed that required parameters have values at the beginning of the method regardless of whether they are nullable.

**Comment by shibayan** [comment]
@mandrean (2017/08), @jimschubert (2017/09) ❤️ @frankyjuang (2019/09) @shibayan (2020/02)

**Comment by wing328** [comment]
@shibayan thanks for fix. Let's go with what you've filed.

**Comment by wing328** [comment]
@shibayan thanks for the PR, which has been included in the v4.3.1 release: https://twitter.com/oas_generator/status/1258057660974329861

**Pull request merged**